jetty-http/src/main/java/org/eclipse/jetty/http/HttpCompliance.java

```diff
@@ -19,99 +19,127 @@
- * <p>
- * Currently the set of modes is an enum and cannot be dynamically extended, but future major releases may convert this
- * to a class. To modify modes there are four custom modes that can be modified by setting the property
- * <code>org.eclipse.jetty.http.HttpCompliance.CUSTOMn</code> (where 'n' is '0', '1', '2' or '3'), to a comma separated
- * list of sections.  The list should start with one of the following strings:<dl>
- * <dt>0</dt><dd>No {@link HttpComplianceSection}s</dd>
- * <dt>*</dt><dd>All {@link HttpComplianceSection}s</dd>
- * <dt>RFC2616</dt><dd>The set of {@link HttpComplianceSection}s application to https://tools.ietf.org/html/rfc2616,
- * but not https://tools.ietf.org/html/rfc7230</dd>
- * <dt>RFC7230</dt><dd>The set of {@link HttpComplianceSection}s application to https://tools.ietf.org/html/rfc7230</dd>
- * </dl>
- * The remainder of the list can contain then names of {@link HttpComplianceSection}s to include them in the mode, or prefixed
- * with a '-' to exclude thm from the mode.    Note that Jetty's modes may have some historic minor differences from the strict
- * RFC compliance, for example the <code>RFC2616_LEGACY</code> HttpCompliance is defined as 
- * <code>RFC2616,-FIELD_COLON,-METHOD_CASE_SENSITIVE</code>.
- * <p>
- * Note also that the {@link EnumSet} return by {@link HttpCompliance#sections()} is mutable, so that modes may 
- * be altered in code and will affect all usages of the mode.
  */
-public enum HttpCompliance // TODO in Jetty-10 convert this enum to a class so that extra custom modes can be defined dynamically
+public final class HttpCompliance
 {
-    /** A Legacy compliance mode to match jetty's behavior prior to RFC2616 and RFC7230. 
-     */
-    LEGACY(sectionsBySpec("0,METHOD_CASE_SENSITIVE")), 
-    
-    /** The legacy RFC2616 support, which incorrectly excludes 
-     * {@link HttpComplianceSection#METHOD_CASE_SENSITIVE}, 
-     * {@link HttpComplianceSection#FIELD_COLON},
-     * {@link HttpComplianceSection#TRANSFER_ENCODING_WITH_CONTENT_LENGTH},
-     * {@link HttpComplianceSection#MULTIPLE_CONTENT_LENGTHS},
-     */
-    RFC2616_LEGACY(sectionsBySpec("RFC2616,-FIELD_COLON,-METHOD_CASE_SENSITIVE,-TRANSFER_ENCODING_WITH_CONTENT_LENGTH,-MULTIPLE_CONTENT_LENGTHS")), 
-    
-    /** The strict RFC2616 support mode */
-    RFC2616(sectionsBySpec("RFC2616")), 
-    
-    /** Jetty's current RFC7230 support, which incorrectly excludes  {@link HttpComplianceSection#METHOD_CASE_SENSITIVE} */
-    RFC7230_LEGACY(sectionsBySpec("RFC7230,-METHOD_CASE_SENSITIVE")),
 
-    /** The RFC7230 support mode */
-    RFC7230(sectionsBySpec("RFC7230"));
+    // These are compliance violations, which may optionally be allowed by the compliance mode, which mean that
+    // the relevant section of the RFC is not strictly adhered to.
+    public enum Violation
+    {
+        CASE_SENSITIVE_FIELD_NAME("https://tools.ietf.org/html/rfc7230#section-3.2", "Field name is case-insensitive"),
```

Comment:
Would really like to see `ALLOW_*` and `REQUIRE_*` prefixes on these.

Reply:
I've gone back and forth on this naming, as I agree that it is kind of attractive to have ALLOW_* and REQUIRE_* in these names.  However, what that does is allow both negative and positive Violations to exist in the enum and would prevent the usage of nice syntax like: `if (CASE_SENSITIVE_FIELD_NAME.isAllowedBy(_compliance))`.      So in the end I think it is best to ensure that all the Violations are actually violations of the RFC/Spec, so that their inclusion in a compliance mode allows us to ask   " if this Violation is allowed by this compliance mode?"